### PR TITLE
Bar element and barchart

### PR DESCRIPTION
**Pull request by JamesxX**
Bar elements are reimplemented with a general solution, which allows explicit control over the placement of each bar in a series. The logic governing clustered and stacked bars can be passed onto the chart's implemented instead.

## The element
The bar element has the following function signature:
```typc
cetz-plot.add.bar(
  data, // Array of unstructured data
  x-key: 0,
  y-key: 1,
  y-base-key: none, // Governs where the bottom of the bar should be drawn (for stacked)
  bar-width: 0.5, // This is the total width of the bar
  label: none,
  style: (:),
  axes: ("x", "y")
)
```

image

Suggest adding support for controlling the number of points along each edge of the bar, for better rendering in other axis styles (for example, more samples on the top horizontal line would allow it to be curved on a polar-2d axis style).

## The chart

Still a work in progress, as I'm taking my time thinking of what people will want to show in a bar chart. I've generally only ever seen bars and error bars, so this is the general case I'll aim to support with the chart (and for more complicated things, it can be constructed manually through plot)

image

## To do:
- [ ] Add support for more samples along straight edges of a bar
- [ ] Put more through into legend preview. It currently shows a half height full width box, so that it will be compatible with the error bar legend preview should it be needed.
- [x] Implement chart
  - [x] Normal
  - [x] Stacked
    - [ ] Colouring of stacked bars is not quite what I'd expect in a chart, order of colours in the legend should match top to bottom the order of colours in a stack.
  - [x] Clustered
  - [x] Stacked100 
- [x] Make sure all bar-related sections in the manual are complete and up to date

### CODE REVIEW

File: src/charts/bar/clustered.typ [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  bar-width: 0.7,
+  bar-spacing: 0,
+  bar-style: palette.red,
+  axes: ("x", "y"),
+  ..plot-args
+) = {
+  let series-count = y-keys.len()
+  bar-width /= series-count
+  let cluster-width = series-count * bar-width + (series-count - 1) * bar-spacing
+  let offsets = (0,)*series-count
```
Comment (johannes-wolf) [inline src/charts/bar/clustered.typ:63]: Unused variable.

File: src/charts/bar/clustered.typ [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+  for (series-index, y-key) in y-keys.enumerate() {
+
+    series-data.push(
+      (
+        label: if label-key != none {labels.at(series-index)},
+        data: for (observation-index, observation) in data.enumerate() { 
+          let x = observation-index - cluster-width/2 + series-index * (bar-width + bar-spacing) + bar-width/2
+          let y = observation.at(y-key, default: 0)
+          offsets.at(observation-index) += y
```
Comment (johannes-wolf) [inline src/charts/bar/clustered.typ:75]: Panics with clustered chart gallery example. Please remove this line as it is unused.

File: src/charts/bar/stacked.typ [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+  y-keys: (1,),
+  y-error-keys: none,
+  bar-width: 0.5,
+  bar-style: palette.red,
+  axes: ("x", "y"),
+  ..plot-args
+) = {
+  let series-count = y-keys.len()
+  let cluster-width = series-count * bar-width
+  let offsets = (0,)*series-count
```
Comment (johannes-wolf) [inline src/charts/bar/stacked.typ:58]: Something is wrong here: `series-count` is the number of `y-keys` but gets accessed with the x-index later which leads to a panic.

File: src/charts/bar/stacked.typ [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+  for (series-index, y-key) in y-keys.enumerate() {
+
+    series-data.push(
+      (
+        label: if label-key != none {labels.at(series-index)},
+        data: for (observation-index, observation) in data.enumerate() { 
+          let x = observation-index
+          let y = observation.at(y-key, default: 0)
+          let y-offset = offsets.at(observation-index)
+          offsets.at(observation-index) += y
```
Comment (johannes-wolf) [inline src/charts/bar/stacked.typ:71]: observation-index can go out of bounds here!

**Review by johannes-wolf — CHANGES_REQUESTED** [review]
The chart API had the option to auto size either the width for bars or the height for columns, that is gone, but I think is useful.

Please add (more) tests for clustered bar-charts.

I've got a fix ready for the non-compiling gallery images that depends on this PR.

File: src/charts/bar/clustered.typ (hunk 2 above)
Comment (JamesxX) [reply src/charts/bar/clustered.typ]: done: [3766d59](https://github.com/cetz-package/cetz-plot/pull/31/commits/3766d59be745136aef9d2322ea4440f0aeecb8a7)

File: src/charts/bar/stacked.typ (hunk 3 above)
Comment (JamesxX) [reply src/charts/bar/stacked.typ]: Resolved by [b0397b3](https://github.com/cetz-package/cetz-plot/pull/31/commits/b0397b316c5410cabfc6d015b7d8823cefae02c2)

File: src/charts/bar/stacked.typ (hunk 4 above)
Comment (JamesxX) [reply src/charts/bar/stacked.typ:71]: Resolved by [b0397b3](https://github.com/cetz-package/cetz-plot/pull/31/commits/b0397b316c5410cabfc6d015b7d8823cefae02c2)

**Comment by JamesxX** [comment]
> The old chart API had the option to auto size either the width for bars or the height for columns, that is gone, but I think is useful.

Autosizing would be great, can we add in a future PR, as it applies to charts more generally than just bar charts - and it would probably depend on the axis flipping we discussed

> Please add (more) tests for clustered bar-charts.

Incoming :)

> I've got a fix ready for the non-compiling gallery images that depends on this PR.
You can also merge that commit into your branch: https://github.com/cetz-package/cetz-plot/pull/32

You know already that I'm a git noob, so suggest we pull that into the aggressive-refactor branch after this is merged

**Comment by JamesxX** [comment]
In a future PR, we may want to add a "errorbar-style" arg but I don't think its essential

**Pull request merged**